### ISSUE TITLE
Systemd-resolved sends incorrect UDP requests to server at 127.0.0.1

### ISSUE THREAD

**Issue by LouHafer** [systemd/systemd#25595]
### systemd version the issue has been seen with

251.8  (systemd-251.8-586.fc37.x86_64)

### Used distribution

Fedora 37

### Linux kernel version used

6.0.9-300.fc37.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd-resolved

### Expected behaviour you didn't see

Resolved should send a UDP DNS request via loopback (lo) to the server listening on the local host at 127.0.0.1

### Unexpected behaviour you saw

Resolved sent a UDP DNS request to 192.168.1.69 (address of the external interface) with destination address 127.0.0.1 via external interface enp7s0 (corresponding external interface). Wireshark says the packet indeed leaves the interface and disappears into the void because no one's listening.

### Steps to reproduce the problem

Standard Fedora 37 install. System with two ethernet interfaces, one to the outside world, one to an internal network. Run a DNS server (BIND 9.18.8) listening at 127.0.0.1. Both ethernet interfaces managed by NetworkManager (1.40.4). Firewalld (1.2.1) is active, configured to masquerade the external interface. External interface uses dhclient (isc-dhclient-4.4.3-P1).  dhclient.conf is
```
interface "enp7s0" {
  prepend domain-search "ivriel" ;
  prepend domain-name-servers 127.0.0.1 ;
}
```
Start systemd-resolved via default systemd configuration. ``resolvectl status`` looks as expected (other DNS servers from my ISP; LLMNR and mDNS disabled in ifcfg-enp7s0)
```
Link 2 (enp7s0)
    Current Scopes: DNS
         Protocols: +DefaultRoute -LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
Current DNS Server: 127.0.0.1
       DNS Servers: 127.0.0.1 192.168.1.254 75.153.171.124 2001:568:ff09:10c::68 2001:568:ff09:10b::124
        DNS Domain: ivriel
```
Recall that enp7s0 has IP address 192.168.1.69 via dhclient. conntrack output shows
```
[NEW] udp      17 30 src=192.168.1.69 dst=127.0.0.1 sport=54484 dport=53 [UNREPLIED] src=127.0.0.1 dst=192.168.1.69 sport=53 dport=54484
```
So I inserted some tracing in nftables (``nft monitor trace`` output attached below for one packet) which also confirmed the packets are headed out enp7s0. Querying the routing tables with ``ip`` always returned ``lo`` as the appropriate interface for a destination of 127.0.0.1.  It struck me I might be able to work around this by forcing a new routing decision. This is the nftables rule labelled ``reroute_OUTPUT`` in the trace below. (Change to DSCP is unimportant in my environment, as far as I'm aware, but triggers a new routing decision.) When the packet arrives at the postrouting hook (``nat_POSTROUTING``) it's headed for the loopback (``lo``) interface. All then works as it should, my local DNS server replies and life is good.

Before I installed the ``reroute_OUTPUT`` rule, resolved would immediately fail over to one of the external DNS servers. If I gave it only my local server to work with (i.e., remove all servers except 127.0.0.1), resolved would eventually degrade to using TCP to connect, which worked just fine. For TCP, the source address was set to 127.0.0.1 (which is what I would expect for both UDP and TCP).

I could convince myself that choosing one of 192.168.1.69 or 127.0.0.1 was policy. But the fact that the choice varies between UDP and TCP queries makes me think this is a bug.

Apologies if this duplicates an existing issue. I did some searching but am probably not sufficiently familiar with the deep network stack terminology to recognise a duplication.

If it's useful, I can go back through the journal and find a set of messages from resolved at debug level from before I installed the ``reroute_OUTPUT`` rule. For what it's worth, they weren't useful to me. The conntrack output was what finally clued me in to the problem.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
trace id b3e6f7f3 inet firewalld trace_OUTPUT packet: oif "enp7s0" ip saddr 192.168.1.69 ip daddr 127.0.0.1 ip dscp cs0 ip ecn not-ect ip ttl 64 ip id 2733 ip protocol udp ip length 72 udp sport 58005 udp dport 53 udp length 52 @th,64,96 0x638001000001000000000001 
trace id b3e6f7f3 inet firewalld trace_OUTPUT rule udp dport 53 meta nftrace set 1 (verdict continue)
trace id b3e6f7f3 inet firewalld trace_OUTPUT verdict continue 
trace id b3e6f7f3 inet firewalld trace_OUTPUT policy accept 
trace id b3e6f7f3 inet firewalld nat_OUTPUT packet: oif "enp7s0" ip saddr 192.168.1.69 ip daddr 127.0.0.1 ip dscp cs0 ip ecn not-ect ip ttl 64 ip id 2733 ip protocol udp ip length 72 udp sport 58005 udp dport 53 udp length 52 @th,64,96 0x638001000001000000000001 
trace id b3e6f7f3 inet firewalld nat_OUTPUT rule jump nat_OUTPUT_POLICIES_pre (verdict jump nat_OUTPUT_POLICIES_pre)
trace id b3e6f7f3 inet firewalld nat_OUTPUT_POLICIES_pre verdict continue 
trace id b3e6f7f3 inet firewalld nat_OUTPUT rule jump nat_OUTPUT_POLICIES_post (verdict jump nat_OUTPUT_POLICIES_post)
trace id b3e6f7f3 inet firewalld nat_OUTPUT_POLICIES_post verdict continue 
trace id b3e6f7f3 inet firewalld nat_OUTPUT verdict continue 
trace id b3e6f7f3 inet firewalld nat_OUTPUT policy accept 
trace id b3e6f7f3 inet firewalld reroute_OUTPUT packet: oif "enp7s0" ip saddr 192.168.1.69 ip daddr 127.0.0.1 ip dscp 0x2a ip ecn not-ect ip ttl 64 ip id 2733 ip protocol udp ip length 72 udp sport 58005 udp dport 53 udp length 52 @th,64,96 0x638001000001000000000001 
trace id b3e6f7f3 inet firewalld reroute_OUTPUT rule oifname "enp7s0" ip daddr 127.0.0.1 udp dport 53 ip dscp set 0x2a (verdict continue)
trace id b3e6f7f3 inet firewalld reroute_OUTPUT verdict continue 
trace id b3e6f7f3 inet firewalld reroute_OUTPUT policy accept 
trace id b3e6f7f3 inet firewalld filter_OUTPUT packet: oif "enp7s0" ip saddr 192.168.1.69 ip daddr 127.0.0.1 ip dscp 0x2a ip ecn not-ect ip ttl 64 ip id 2733 ip protocol udp ip length 72 udp sport 58005 udp dport 53 udp length 52 @th,64,96 0x638001000001000000000001 
trace id b3e6f7f3 inet firewalld filter_OUTPUT rule jump filter_OUTPUT_POLICIES_pre (verdict jump filter_OUTPUT_POLICIES_pre)
trace id b3e6f7f3 inet firewalld filter_OUTPUT_POLICIES_pre verdict continue 
trace id b3e6f7f3 inet firewalld filter_OUTPUT rule jump filter_OUTPUT_POLICIES_post (verdict jump filter_OUTPUT_POLICIES_post)
trace id b3e6f7f3 inet firewalld filter_OUTPUT_POLICIES_post verdict continue 
trace id b3e6f7f3 inet firewalld filter_OUTPUT verdict continue 
trace id b3e6f7f3 inet firewalld filter_OUTPUT policy accept 
trace id b3e6f7f3 inet firewalld nat_POSTROUTING packet: oif "lo" ip saddr 192.168.1.69 ip daddr 127.0.0.1 ip dscp 0x2a ip ecn not-ect ip ttl 64 ip id 2733 ip protocol udp ip length 72 udp sport 58005 udp dport 53 udp length 52 @th,64,96 0x638001000001000000000001 
trace id b3e6f7f3 inet firewalld nat_POSTROUTING rule jump nat_POSTROUTING_ZONES (verdict jump nat_POSTROUTING_ZONES)
trace id b3e6f7f3 inet firewalld nat_POSTROUTING_ZONES rule goto nat_POST_drop (verdict goto nat_POST_drop)
trace id b3e6f7f3 inet firewalld nat_POST_drop rule jump nat_POSTROUTING_POLICIES_pre (verdict jump nat_POSTROUTING_POLICIES_pre)
trace id b3e6f7f3 inet firewalld nat_POSTROUTING_POLICIES_pre verdict continue 
trace id b3e6f7f3 inet firewalld nat_POST_drop rule jump nat_POST_drop_pre (verdict jump nat_POST_drop_pre)
trace id b3e6f7f3 inet firewalld nat_POST_drop_pre verdict continue 
trace id b3e6f7f3 inet firewalld nat_POST_drop rule jump nat_POST_drop_log (verdict jump nat_POST_drop_log)
trace id b3e6f7f3 inet firewalld nat_POST_drop_log verdict continue 
trace id b3e6f7f3 inet firewalld nat_POST_drop rule jump nat_POST_drop_deny (verdict jump nat_POST_drop_deny)
trace id b3e6f7f3 inet firewalld nat_POST_drop_deny verdict continue 
trace id b3e6f7f3 inet firewalld nat_POST_drop rule jump nat_POST_drop_allow (verdict jump nat_POST_drop_allow)
trace id b3e6f7f3 inet firewalld nat_POST_drop_allow verdict continue 
trace id b3e6f7f3 inet firewalld nat_POST_drop rule jump nat_POST_drop_post (verdict jump nat_POST_drop_post)
trace id b3e6f7f3 inet firewalld nat_POST_drop_post verdict continue 
trace id b3e6f7f3 inet firewalld nat_POST_drop rule jump nat_POSTROUTING_POLICIES_post (verdict jump nat_POSTROUTING_POLICIES_post)
trace id b3e6f7f3 inet firewalld nat_POSTROUTING_POLICIES_post verdict continue 
trace id b3e6f7f3 inet firewalld nat_POST_drop verdict continue 
trace id b3e6f7f3 inet firewalld nat_POSTROUTING verdict continue 
trace id b3e6f7f3 inet firewalld nat_POSTROUTING policy accept
```

**Comment by yuwata** [comment]
This should be already fixed by 6e32414a66ff8dbcef233981a7066684d903ee9f. Closing.
Feel free to request to reopen this page if the commit does not fix your issue.

**Comment by LouHafer** [comment]
Duh. Should have thought to check closed issues. Thanks for the quick response. Hopefully it'll propagate quickly through the distribution channels.
